frame/support/src/storage/types/key.rs

```diff
@@ -0,0 +1,900 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage key type.
+
+use crate::hash::{ReversibleStorageHasher, StorageHasher};
+use codec::{Encode, EncodeLike, FullCodec};
+use paste::paste;
+use sp_std::prelude::*;
+
+/// A type used exclusively by storage maps as their key type.
+///
+/// The final key generated has the following form:
+/// ```nocompile
+/// Hasher1(encode(key1))
+///     ++ Hasher2(encode(key2))
+///     ++ ...
+///     ++ HasherN(encode(keyN))
+/// ```
+pub struct Key<Hasher, KeyType>(core::marker::PhantomData<(Hasher, KeyType)>);
+
+/// A trait that contains the current key as an associated type.
+pub trait KeyGenerator {
+    type Key: EncodeLike<Self::Key>;
+    type KArg: Encode;
+    type HashFn: FnOnce(&[u8]) -> Vec<u8>;
+    type HArg;
+
+    fn final_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8>;
+    fn migrate_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(
+        key: &KArg,
+        hash_fns: Self::HArg,
+    ) -> Vec<u8>;
+    fn final_hash(encoded: &[u8]) -> Vec<u8>;
+}
+
+impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
+    type Key = K;
+    type KArg = (K,);
+    type HashFn = Box<dyn FnOnce(&[u8]) -> Vec<u8>>;
+    type HArg = (Self::HashFn,);
+
+    fn final_key<KArg: EncodeLike<Self::KArg> + TupleToEncodedIter>(key: KArg) -> Vec<u8> {
+        H::hash(
+            &key.to_encoded_iter()
+                .next()
+                .expect("should have at least one element!"),
```

Comment:
I'm not sure the trait bounds enforce this.

`A: EncodeLike<B>` means means A encode likes B: i.e. an encoding of a value of type A results in a valid encoding for the type B (it can be decoded in type B).

I know the implementation of EncodeLike for tuples currently enforce that if A and B are tuples, then it contains the same number of element. But actually nothing prevents from implementing encode like for some tuples which would break this assumption.
Actually we could write: `impl<A: Encode> EncodeLike<(A, ())> for (A,) {}`.
I consider it to be a valid implementation of `EncodeLike` (e.g. `(3u32,).encode()` can be decoded in `(u32, ())`).
But the 2 tuple types have a different number of element.

To solve this I think we should create a trait which ensure `KArg` and `Self::KArg` are both tuples with the same number of element and each element encodes like their corresponding ones.

(note: maybe this trait can replace `TupleToEncodedIter` at the same time.)

Reply:
This one is a tad bit difficult, I've spent quite a majority of time trying to get things to work with `TupleToEncodedIter`, and I couldn't quite figure out a better way to 1) be generic and accept tuple parameters with varying arities and 2) loop through each element in the tuple and extract the encoded vector out of them.

I suppose what I can do is create a `EncodeLikeTuple<T>` trait and implement it on various tuples with different arities, e.g. `impl EncodeLikeTuple<(A,B)> for (AA, BB)` and so on, and perhaps then define a trait method that allows me to send in a hasher and encode the elements accordingly. I'd probably need to construct a new macro to generate these implementations; I don't think I can use `impl-trait-for-tuples` here because of the type parameters to `EncodeLikeTuple`.

Reply:
yes as far as I can see the easiest implementation is another trait EncodeLikeTuple which ensure that every element of some tuple encode likes their corresponding element on the other tuple (and thus ensuring both tuple have same number of element)

You can copy/paste what we did in parity-scale-codec if wanted: https://github.com/paritytech/parity-scale-codec/blob/f0341dabb01aa9ff0548558abb6dcc5c31c669a1/src/codec.rs#L1002-L1100

but I'm not sure I understand what you mean by `"and perhaps then define a trait method that allows me to send in a hasher and encode the elements accordingly"` for me the current implementation is fine, just the trait needs to bound more assumption on the types. Once you replace `KArg: EncodeLike<Self::KArg> + TupleToEncodedIter` by `KArg: EncodeLikeTuple<Self::KArg>` the implementation is correct and making correct assumption about the types.

do I miss something ?

Reply:
The entire reason why I had to create the `TupleToEncodedIter` trait and add it as a trait bound to `KArg` is because I'm unable to do any tuple indexing on `KArg`, since there is no type information indicating that `KArg` is actually a tuple underneath.

And so by keeping in mind the ultimate goal here is to encode and hash each tuple element accordingly, maybe I'll have to create a trait function, something like `fn encode_and_hash<H: Hasher>(&self) -> Vec<u8>`, and call it on each individual tuple element.

But this brings up the problem of knowing how to obtain the `StorageHasher` used for that particular key in the tuple. It's not a straightforward answer, because it's defined on the `Key` struct. I think I might have a way if I define the trait bounds appropriately, so I'll see how much I can push this to work.

Reply:
> Actually, the toughest challenge here is obtaining each tuple element from KArg.

But you accomplished this with `TupleToEncodeIter`. the only rememaining stuff to accomplish is to ensure this `expect` is true for any correct implementation of traits. This can already be done by copying EncodeLike trait and its implementation on tuple. Just renaming it to something like EncodeLikeTuple and say in the doc that it ensures that tuple implements encode like tuple for which each of its element encode like the element of the other tuple.

I'll read your message again to understand the struggle but I don't see why it is not possible as the current implementation already works. We just need to ensure that tuple cannot implement EncodeLike to another tuple if each of its element doesn't encode like the element of the other tuple.

We create a new trait `EncodeLikeTuple<T>` implement it similarly like EncodeLike for the tuples.

Reply:
> but getting the index for each KArg is near impossible, since self.Tuple in the for_tuple macro doesn't do the job, and I can't do key.Tuple either.

then don't use for_tuple macro to implement EncodeLikeTuple, the macro in parity-scale-codec should be fine

Reply:
I'm ok to have 2 different trait TupleToEncodeIter and EncodeLikeTuple if needed

Reply:
yes looks good to me now